src/clr-addons/datagrid/datagrid-state-persistence/state-persistence-key.directive.ts

```diff
@@ -1,35 +1,156 @@
-      this.pagination.page.sizeChange.subscribe(pageSize => {
-        let state = JSON.parse(localStorage.getItem(this.clrStatePersistenceKey)) as ClrDatagridStatePersistenceModel;
-        if (!state) {
-          state = {} as ClrDatagridStatePersistenceModel;
+      this.initPageSizePersister(localStorageState);
+      this.initCurrentPage(sessionState);
+    }
+  }
+
+  private initPageSizePersister(savedState: ClrDatagridStatePersistenceModel) {
+    /* persist page size changes in local storage */
+    this.pagination.page.sizeChange.pipe(takeUntil(this.destroy$)).subscribe(pageSize => {
+      const state = this.getLocalStorageState();
+      state.pageSize = pageSize;
+      localStorage.setItem(this.options.key, JSON.stringify(state));
+    });
+
+    /* init page size of datagrid if already persisted in local storage */
+    if (savedState.pageSize) {
+      this.pagination.page.size = savedState.pageSize;
+    }
+  }
+
+  private initCurrentPage(savedState: ClrDatagridStatePersistenceModel) {
+    /* init current page of datagrid if already persisted */
+    if (savedState?.currentPage) {
+      this.pagination.page.current = savedState.currentPage;
+    }
+  }
+
+  private initFilter(savedState: ClrDatagridStatePersistenceModel) {
+    if (savedState.columns) {
+      Object.keys(savedState.columns).forEach(prop => {
+        const filter = this.getFilter(prop);
+        if (filter) {
+          const savedFilterValue = savedState.columns[prop].filterValue || {};
+          Object.keys(savedFilterValue)
+            .filter(valueKey => valueKey !== 'property')
+            .forEach(valueKey => ((filter as any)[valueKey] = this.parseFilterValue(savedFilterValue[valueKey])));
         }
-        state.pageSize = pageSize;
-        localStorage.setItem(this.clrStatePersistenceKey, JSON.stringify(state));
       });
-
-      /* init page size of datagrid if already persisted in local storage */
-      const loadedState = JSON.parse(localStorage.getItem(this.clrStatePersistenceKey));
-      if (loadedState?.pageSize) {
-        /* postpone set size to other cycle as it is already set in this change detection cycle */
-        setTimeout(() => (this.pagination.page.size = loadedState.pageSize));
-      }
     }
   }
+
+  private initDatagridPersister() {
+    this.datagrid.refresh.pipe(takeUntil(this.destroy$)).subscribe(dgState => {
+      const state = this.getSessionState();
+
+      state.currentPage = dgState.page?.current;
+      state.columns = state.columns || {};
+      Object.keys(state.columns).forEach(prop => (state.columns[prop].filterValue = undefined));
+      dgState.filters?.forEach(filter => {
+        const property = this.getFilterPropertyName(filter);
+        if (property) {
+          state.columns[property] = state.columns[property] || {};
+          state.columns[property].filterValue = this.enrichFilterValue(filter);
+        }
+      });
+
+      sessionStorage.setItem(this.options.key, JSON.stringify(state));
```

Comment:
see Teams message about `LocalStorage` vs `SessionStorage`

Reply:
volatile values (current page and filters) are stored in SessionStorage to not have the need of a reset button.
more static values (hidden state and page size) are stored in LocalStorage as they should almost always stay the same.

Reply:
option added, resolved